longitudinal_controller/src/longitudinal_controller.py

```diff
@@ -0,0 +1,227 @@
+#!/usr/bin/python3
+
+from enum import Enum
+import numpy as np
+import rospy
+import time
+
+from pid import PID
+from geometry_msgs.msg import TwistStamped
+from self_racing_car_msgs.msg import ArduinoLogging
+from std_msgs.msg import Float32, Float64MultiArray
+
+
+class LongitudinalControlMode(Enum):
+    ConstantPwmOutput = 1
+    PID = 2
+
+
+class LongitudinalController:
+    def __init__(self):
+
+        # Constants
+        rate = rospy.get_param("~rate", 10.0)
+        self.throttle_idle_autonomous_pwm = rospy.get_param(
+            "~throttle_idle_autonomous_pwm", 90
+        )
+        self.throttle_max_autonomous_pwm = rospy.get_param(
+            "~throttle_max_autonomous_pwm", 102
+        )
+        self.throttle_min_autonomous_pwm = rospy.get_param(
+            "~throttle_min_autonomous_pwm", 70
+        )
+        self.longitudinal_control_mode = rospy.get_param(
+            "~longitudinal_control_mode", 2
+        )
+        gain_p = rospy.get_param("~speed_control_gain_p", 0.5)
+        gain_i = rospy.get_param("~speed_control_gain_i", 0.1)
+        gain_d = rospy.get_param("~speed_control_gain_d", 0)
+        self.constant_pwm_output = rospy.get_param(
+            "~constant_pwm_output", 90
+        )  # If ConstantPwmOutput mode
+        self.timeout_engage_msg_before_stop_secs = rospy.get_param(
+            "~timeout_engage_msg_before_stop_secs", 1
+        )
+
+        # Subscribers
+        rospy.Subscriber(
+            "target_velocity",
+            TwistStamped,
+            self.target_velocity_callback,
+        )
+        rospy.Subscriber(
+            "current_velocity",
+            TwistStamped,
+            self.current_velocity_callback,
+        )
+        rospy.Subscriber(
+            "arduino_logging",
+            ArduinoLogging,
+            self.firmware_log_callback,
+        )
+
+        # Publishers
+        self.throttle_cmd_pub = rospy.Publisher(
+            "throttle_pwm_cmd",
+            Float32,
+            queue_size=10,
+        )
+
+        self.pub_debug_pid = rospy.Publisher(
+            "pid_debug", Float64MultiArray, queue_size=10
+        )
+
+        # Variables
+        self.desired_velocity = 0
+        self.current_velocity = 0
+        self.anti_windup_enabled = False
+        self.pid_controller = PID(gain_p, gain_i, gain_d)
+        self.speed_controller_enabled = False
+        self.previous_t = time.time()
+        self.rate = rospy.Rate(rate)
+        self.time_last_enable_msg = time.time()
+
+    def current_velocity_callback(self, msg: TwistStamped):
+        """
+        Callback function for current velocity message
+        """
+        self.current_velocity = msg.twist.linear.x
+
+    def firmware_log_callback(self, msg: ArduinoLogging):
+        """
+        Callback of log message from low level controller.
+        """
+        self.speed_controller_enabled = msg.engaged_mode
+        self.time_last_enable_msg = time.time()
```

Comment:
Don't we need to check whether we are in engaged mode here before setting  `self.time_last_enable_msg` to `time.time()`

Reply:
I think what I had in mind here is have a way to know if self.speed_controller_enabled has been updated recently, whether or not it is equal to True or False, that's why I reset the time ever time we enter in the callback.
But I guess we could add the logic you describe here. What do you think?

Reply:
Ok that makes sense. For better clarity, we could rename ` self.time_last_enable_msg` into something like ` self.time_last_arduino_msg` (though we loose a bit of generality in the naming by saying it's from the Arduino...)

Reply:
Done